### PR TITLE
[feat]: checkpoint and normalization

### PR DESCRIPTION
**Pull request by min-xu-ai**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/master/CONTRIBUTING.md)?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

## What does this PR do?
Fixes # (issue).

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/nn/misc/test_checkpoint_activations_norm.py
Comment (min-xu-ai) [inline tests/nn/misc/test_checkpoint_activations_norm.py:74]: these are probably to be expected as half not implemented on cpu for both LN and BN.

File: tests/nn/misc/test_checkpoint_activations_norm.py
Comment (min-xu-ai) [inline tests/nn/misc/test_checkpoint_activations_norm.py:79]: this is interesting, if we just change param and buffer dtype, not call half(), it fails with this error on CUDA for BN.

File: fairscale/nn/misc/checkpoint_activations.py
Comment (myleott) [inline fairscale/nn/misc/checkpoint_activations.py:93]: can we directly `register_forward_pre_hook` and `register_forward_hook` on the module instead of piping these through the checkpointing method?

File: fairscale/nn/misc/checkpoint_activations.py
Comment (myleott) [inline fairscale/nn/misc/checkpoint_activations.py:83]: This actually seems like a general problem, I wonder if we can make something more standalone/reusable? In particular, I feel that BatchNorm layers should not update stats when in `torch.no_grad()` mode. Currently they only respect `model.eval()`, but perhaps we can have a simple module-level wrapper that registers the relevant hooks?

Something like (untested):
```python
def patch_batch_norm(module: nn.Module):
    if isinstance(module, _BatchNorm):
        module.register_forward_pre_hook(bn_pre_hook)
        module.register_forward_hook(bn_post_hook)
    return module

def bn_pre_hook(module, input):
    if not torch.is_grad_enabled():
        module._backup_track_running_stats = module.track_running_stats
        module.track_running_stats = False

def bn_post_hook(module, input, output):
    if not torch.is_grad_enabled():
        module.track_running_stats = module._backup_track_running_stats
```

What do you think?

**Comment by anj-s** [comment]
@min-xu-ai Can you give me some context about what the feature is doing? Is it for supporting BN?

File: fairscale/nn/misc/checkpoint_activations.py
Comment (min-xu-ai) [reply fairscale/nn/misc/checkpoint_activations.py:83]: Great idea! Thanks for suggesting. Let me give it a try. I am a bit worries that users may want to keep the behavior of tracking stats under no_grad() but that's probably a very rare thing if at all possible. The current method only affect checkpointed fwd pass.

**Comment by min-xu-ai** [comment]
> @min-xu-ai Can you give me some context about what the feature is doing? Is it for supporting BN?

Thanks for taking a look, @anj-s!

Yes, this change stems from that we observed that something is off for BN under activation checkpointing (not just the version in fairscale, also under the version in pytorch). This is an attempt to workaround that.

We also observed some differences in how to we fp16 for BN affects its behavior. So the unit test includes that as well.

File: tests/nn/misc/test_checkpoint_activations_norm.py
Comment (min-xu-ai) [reply tests/nn/misc/test_checkpoint_activations_norm.py:79]: @myleott What do you think about this? I guess we should workaround that too in BN?

Comment (min-xu-ai) [reply tests/nn/misc/test_checkpoint_activations_norm.py:79]: cc @sshleifer too.

**Comment by min-xu-ai** [comment]
@myleott @anj-s , I think this is ready. Mind taking another look? Note, the benchmark CI tests are currently failing due to a networking error. Will have to deal with them separately.

File: tests/nn/misc/test_checkpoint_activations_norm.py
Comment (myleott) [inline tests/nn/misc/test_checkpoint_activations_norm.py:39]: not: make this `elif`?

File: tests/nn/misc/test_checkpoint_activations_norm.py
Comment (myleott) [inline tests/nn/misc/test_checkpoint_activations_norm.py:37]: shouldn't this be `setattr(m, n, b.half())`?

File: tests/nn/misc/test_checkpoint_activations_norm.py
Comment (myleott) [reply tests/nn/misc/test_checkpoint_activations_norm.py:79]: I think this is a bug in the test (see comment above)

**Review by myleott — APPROVED** [review]
Nice! Small fix needed, otherwise LGTM

File: tests/nn/misc/test_checkpoint_activations_norm.py
Comment (min-xu-ai) [reply tests/nn/misc/test_checkpoint_activations_norm.py:37]: yes! what's a dumb bug!

**Comment by min-xu-ai** [comment]
test bug fixed and addressed comments. Will merge if everything passes.

**Pull request merged**
